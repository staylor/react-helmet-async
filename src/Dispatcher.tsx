import { Component } from 'react';
import shallowEqual from 'shallowequal';

import handleStateChangeOnClient from './client';
import mapStateOnServer from './server';
import { reducePropsToState } from './utils';
import Provider from './Provider';
import type { HelmetServerState } from './types';

export interface DispatcherContextProp {
  setHelmet: (newState: HelmetServerState) => void;
  helmetInstances: {
    get: () => HelmetDispatcher[];
    add: (helmet: HelmetDispatcher) => void;
    remove: (helmet: HelmetDispatcher) => void;
  };
}

interface DispatcherProps {
  context: DispatcherContextProp;
}

export default class HelmetDispatcher extends Component<DispatcherProps> {
  rendered = false;

  shouldComponentUpdate(nextProps: DispatcherProps) {
    return !shallowEqual(nextProps, this.props);
  }

  emitChange() {
    const { helmetInstances, setHelmet } = this.props.context;
    let serverState = null;
    const state = reducePropsToState(
      helmetInstances.get().map(instance => {
        const props = { ...instance.props };
        // @ts-ignore
        delete props.context;
        return props;
      })
    );
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      serverState = mapStateOnServer(state);
    }
    // @ts-ignore
    setHelmet(serverState);
  }

  // componentWillMount will be deprecated
  // for SSR, initialize on first render
  // constructor is also unsafe in StrictMode
  init() {
    if (this.rendered) {
      return;
    }

    this.rendered = true;

    const { helmetInstances } = this.props.context;
    helmetInstances.add(this);
    this.emitChange();
  }

  render() {
    this.init();

    return null;
  }
}
