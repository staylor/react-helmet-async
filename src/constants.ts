export enum TAG_PROPERTIES {
  CHARSET = 'charset',
  CSS_TEXT = 'cssText',
  HREF = 'href',
  HTTPEQUIV = 'http-equiv',
  INNER_HTML = 'innerHTML',
  ITEM_PROP = 'itemprop',
  NAME = 'name',
  PROPERTY = 'property',
  REL = 'rel',
  SRC = 'src',
}

export enum ATTRIBUTE_NAMES {
  BODY = 'bodyAttributes',
  HTML = 'htmlAttributes',
  TITLE = 'titleAttributes',
}

export enum TAG_NAMES {
  BASE = 'base',
  BODY = 'body',
  HEAD = 'head',
  HTML = 'html',
  LINK = 'link',
  META = 'meta',
  NOSCRIPT = 'noscript',
  SCRIPT = 'script',
  STYLE = 'style',
  TITLE = 'title',
  FRAGMENT = 'Symbol(react.fragment)',
}

export const SEO_PRIORITY_TAGS = {
  link: { rel: ['amphtml', 'canonical', 'alternate'] },
  script: { type: ['application/ld+json'] },
  meta: {
    charset: '',
    name: ['generator', 'robots', 'description'],
    property: [
      'og:type',
      'og:title',
      'og:url',
      'og:image',
      'og:image:alt',
      'og:description',
      'twitter:url',
      'twitter:title',
      'twitter:description',
      'twitter:image',
      'twitter:image:alt',
      'twitter:card',
      'twitter:site',
    ],
  },
};

export const VALID_TAG_NAMES = Object.values(TAG_NAMES);

export const REACT_TAG_MAP = {
  accesskey: 'accessKey',
  charset: 'charSet',
  class: 'className',
  contenteditable: 'contentEditable',
  contextmenu: 'contextMenu',
  'http-equiv': 'httpEquiv',
  itemprop: 'itemProp',
  tabindex: 'tabIndex',
};

export const HTML_TAG_MAP = Object.entries(REACT_TAG_MAP).reduce(
  (carry, [key, value]) => {
    carry[value] = key;
    return carry;
  },
  {} as { [key: string]: string }
);

export const HELMET_ATTRIBUTE = 'data-rh';
