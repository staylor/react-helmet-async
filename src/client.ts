import { HELMET_ATTRIBUTE, TAG_EVENTS, TAG_NAMES, TAG_PROPERTIES } from './constants';
import type { Attributes, StateUpdate, TagList } from './types';
import { flattenArray } from './utils';

interface TagUpdates {
  oldTags: HTMLElement[];
  newTags: HTMLElement[];
}

interface TagUpdateList {
  [key: string]: TagUpdates;
}

const updateTags = (type: string, tags: HTMLElement[]) => {
  const headElement = document.head || document.querySelector(TAG_NAMES.HEAD);
  const tagNodes = headElement.querySelectorAll(`${type}[${HELMET_ATTRIBUTE}]`);
  const oldTags: HTMLElement[] = [].slice.call(tagNodes);
  const newTags: HTMLElement[] = [];
  let indexToDelete: number;

  if (tags && tags.length) {
    tags.forEach(tag => {
      const newElement = document.createElement(type);

      for (const attribute in tag) {
        if (Object.prototype.hasOwnProperty.call(tag, attribute)) {
          if (attribute === TAG_PROPERTIES.INNER_HTML) {
            newElement.innerHTML = tag.innerHTML;
          } else if (attribute === TAG_PROPERTIES.CSS_TEXT) {
            // This seems like a CSSImportRuleDeclaration?
            // @ts-ignore
            if (newElement.styleSheet) {
              // @ts-ignore
              newElement.styleSheet.cssText = tag.cssText;
            } else {
              // @ts-ignore
              newElement.appendChild(document.createTextNode(tag.cssText));
            }
          } else if (
            (TAG_EVENTS.ON_LOAD === attribute || TAG_EVENTS.ON_ERROR === attribute) &&
            typeof tag[attribute] === 'function' &&
            newElement instanceof HTMLScriptElement
          ) {
            const attr = attribute as keyof HTMLElement;
            const value = typeof tag[attr] === 'undefined' ? '' : tag[attr];
            newElement[attribute] = value as any;
          } else {
            const attr = attribute as keyof HTMLElement;
            const value = typeof tag[attr] === 'undefined' ? '' : tag[attr];
            newElement.setAttribute(attribute, value as string);
          }
        }
      }

      newElement.setAttribute(HELMET_ATTRIBUTE, 'true');

      // Remove a duplicate tag from domTagstoRemove, so it isn't cleared.
      if (
        oldTags.some((existingTag, index) => {
          indexToDelete = index;
          return newElement.isEqualNode(existingTag);
        })
      ) {
        oldTags.splice(indexToDelete, 1);
      } else {
        newTags.push(newElement);
      }
    });
  }

  oldTags.forEach((tag: Node) => tag.parentNode?.removeChild(tag));
  newTags.forEach(tag => headElement.appendChild(tag));

  return {
    oldTags,
    newTags,
  };
};

const updateAttributes = (tagName: string, attributes: Attributes) => {
  const elementTag = document.getElementsByTagName(tagName)[0];

  if (!elementTag) {
    return;
  }

  const helmetAttributeString = elementTag.getAttribute(HELMET_ATTRIBUTE);
  const helmetAttributes = helmetAttributeString ? helmetAttributeString.split(',') : [];
  const attributesToRemove = [...helmetAttributes];
  const attributeKeys = Object.keys(attributes);

  for (const attribute of attributeKeys) {
    const value = attributes[attribute] || '';

    if (elementTag.getAttribute(attribute) !== value) {
      elementTag.setAttribute(attribute, value);
    }

    if (helmetAttributes.indexOf(attribute) === -1) {
      helmetAttributes.push(attribute);
    }

    const indexToSave = attributesToRemove.indexOf(attribute);
    if (indexToSave !== -1) {
      attributesToRemove.splice(indexToSave, 1);
    }
  }

  for (let i = attributesToRemove.length - 1; i >= 0; i -= 1) {
    elementTag.removeAttribute(attributesToRemove[i]);
  }

  if (helmetAttributes.length === attributesToRemove.length) {
    elementTag.removeAttribute(HELMET_ATTRIBUTE);
  } else if (elementTag.getAttribute(HELMET_ATTRIBUTE) !== attributeKeys.join(',')) {
    elementTag.setAttribute(HELMET_ATTRIBUTE, attributeKeys.join(','));
  }
};

const updateTitle = (title: string, attributes: Attributes) => {
  if (typeof title !== 'undefined' && document.title !== title) {
    document.title = flattenArray(title);
  }

  updateAttributes(TAG_NAMES.TITLE, attributes);
};

type Cb = () => number | null | void;

const commitTagChanges = (newState: StateUpdate, cb?: Cb) => {
  const {
    baseTag,
    bodyAttributes,
    htmlAttributes,
    linkTags,
    metaTags,
    noscriptTags,
    onChangeClientState,
    scriptTags,
    styleTags,
    title,
    titleAttributes,
  } = newState;
  updateAttributes(TAG_NAMES.BODY, bodyAttributes as Attributes);
  updateAttributes(TAG_NAMES.HTML, htmlAttributes as Attributes);

  updateTitle(title, titleAttributes as Attributes);

  const tagUpdates: TagUpdateList = {
    baseTag: updateTags(TAG_NAMES.BASE, baseTag),
    linkTags: updateTags(TAG_NAMES.LINK, linkTags),
    metaTags: updateTags(TAG_NAMES.META, metaTags),
    noscriptTags: updateTags(TAG_NAMES.NOSCRIPT, noscriptTags),
    scriptTags: updateTags(TAG_NAMES.SCRIPT, scriptTags),
    styleTags: updateTags(TAG_NAMES.STYLE, styleTags),
  };

  const addedTags: TagList = {};
  const removedTags: TagList = {};

  Object.keys(tagUpdates).forEach(tagType => {
    const { newTags, oldTags } = tagUpdates[tagType];

    if (newTags.length) {
      addedTags[tagType] = newTags;
    }
    if (oldTags.length) {
      removedTags[tagType] = tagUpdates[tagType].oldTags;
    }
  });

  if (cb) {
    cb();
  }

  onChangeClientState(newState, addedTags, removedTags);
};

let _helmetCallback: number | null = null;

const handleStateChangeOnClient = (newState: StateUpdate) => {
  if (_helmetCallback) {
    cancelAnimationFrame(_helmetCallback);
  }

  if (newState.defer) {
    _helmetCallback = requestAnimationFrame(() => {
      commitTagChanges(newState, () => {
        _helmetCallback = null;
      });
    });
  } else {
    commitTagChanges(newState);
    _helmetCallback = null;
  }
};

export default handleStateChangeOnClient;
