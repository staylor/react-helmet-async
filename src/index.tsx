import type { PropsWithChildren, ReactElement, ReactNode } from 'react';
import React, { Component } from 'react';
import fastCompare from 'react-fast-compare';
import invariant from 'invariant';

import { Context } from './Provider';
import type { HelmetDataType } from './HelmetData';
import HelmetData from './HelmetData';
import type { DispatcherContextProp } from './Dispatcher';
import Dispatcher from './Dispatcher';
import { without } from './utils';
import { TAG_NAMES, VALID_TAG_NAMES, HTML_TAG_MAP } from './constants';
import type { HelmetProps } from './types';

export * from './types';

export { default as HelmetData } from './HelmetData';
export { default as HelmetProvider } from './Provider';

type Props = { [key: string]: any };

export class Helmet extends Component<PropsWithChildren<HelmetProps>> {
  static defaultProps = {
    defer: true,
    encodeSpecialCharacters: true,
    prioritizeSeoTags: false,
  };

  shouldComponentUpdate(nextProps: HelmetProps) {
    return !fastCompare(without(this.props, 'helmetData'), without(nextProps, 'helmetData'));
  }

  mapNestedChildrenToProps(child: ReactElement, nestedChildren: ReactNode) {
    if (!nestedChildren) {
      return null;
    }

    switch (child.type) {
      case TAG_NAMES.SCRIPT:
      case TAG_NAMES.NOSCRIPT:
        return {
          innerHTML: nestedChildren,
        };

      case TAG_NAMES.STYLE:
        return {
          cssText: nestedChildren,
        };
      default:
        throw new Error(
          `<${child.type} /> elements are self-closing and can not contain children. Refer to our API for more information.`
        );
    }
  }

  flattenArrayTypeChildren(
    child: JSX.Element,
    arrayTypeChildren: { [key: string]: JSX.Element[] },
    newChildProps: Props,
    nestedChildren: ReactNode
  ) {
    return {
      ...arrayTypeChildren,
      [child.type]: [
        ...(arrayTypeChildren[child.type] || []),
        {
          ...newChildProps,
          ...this.mapNestedChildrenToProps(child, nestedChildren),
        },
      ],
    };
  }

  mapObjectTypeChildren(
    child: JSX.Element,
    newProps: Props,
    newChildProps: Props,
    nestedChildren: ReactNode
  ) {
    switch (child.type) {
      case TAG_NAMES.TITLE:
        return {
          ...newProps,
          [child.type]: nestedChildren,
          titleAttributes: { ...newChildProps },
        };

      case TAG_NAMES.BODY:
        return {
          ...newProps,
          bodyAttributes: { ...newChildProps },
        };

      case TAG_NAMES.HTML:
        return {
          ...newProps,
          htmlAttributes: { ...newChildProps },
        };
      default:
        return {
          ...newProps,
          [child.type]: { ...newChildProps },
        };
    }
  }

  mapArrayTypeChildrenToProps(arrayTypeChildren: { [key: string]: JSX.Element }, newProps: Props) {
    let newFlattenedProps = { ...newProps };

    Object.keys(arrayTypeChildren).forEach(arrayChildName => {
      newFlattenedProps = {
        ...newFlattenedProps,
        [arrayChildName]: arrayTypeChildren[arrayChildName],
      };
    });

    return newFlattenedProps;
  }

  warnOnInvalidChildren(child: JSX.Element, nestedChildren: ReactNode) {
    invariant(
      VALID_TAG_NAMES.some(name => child.type === name),
      typeof child.type === 'function'
        ? `You may be attempting to nest <Helmet> components within each other, which is not allowed. Refer to our API for more information.`
        : `Only elements types ${VALID_TAG_NAMES.join(
            ', '
          )} are allowed. Helmet does not support rendering <${
            child.type
          }> elements. Refer to our API for more information.`
    );

    invariant(
      !nestedChildren ||
        typeof nestedChildren === 'string' ||
        (Array.isArray(nestedChildren) &&
          !nestedChildren.some(nestedChild => typeof nestedChild !== 'string')),
      `Helmet expects a string as a child of <${child.type}>. Did you forget to wrap your children in braces? ( <${child.type}>{\`\`}</${child.type}> ) Refer to our API for more information.`
    );

    return true;
  }

  mapChildrenToProps(children: ReactNode, newProps: Props) {
    let arrayTypeChildren = {};

    React.Children.forEach(children as JSX.Element, (child: ReactElement) => {
      if (!child || !child.props) {
        return;
      }

      const { children: nestedChildren, ...childProps } = child.props;
      // convert React props to HTML attributes
      const newChildProps = Object.keys(childProps).reduce((obj: Props, key) => {
        obj[HTML_TAG_MAP[key] || key] = childProps[key];
        return obj;
      }, {});

      let { type } = child;
      if (typeof type === 'symbol') {
        type = (type as 'symbol').toString();
      } else {
        this.warnOnInvalidChildren(child, nestedChildren);
      }

      switch (type) {
        case TAG_NAMES.FRAGMENT:
          newProps = this.mapChildrenToProps(nestedChildren, newProps);
          break;

        case TAG_NAMES.LINK:
        case TAG_NAMES.META:
        case TAG_NAMES.NOSCRIPT:
        case TAG_NAMES.SCRIPT:
        case TAG_NAMES.STYLE:
          arrayTypeChildren = this.flattenArrayTypeChildren(
            child,
            arrayTypeChildren,
            newChildProps,
            nestedChildren
          );
          break;

        default:
          newProps = this.mapObjectTypeChildren(child, newProps, newChildProps, nestedChildren);
          break;
      }
    });

    return this.mapArrayTypeChildrenToProps(arrayTypeChildren, newProps);
  }

  render() {
    const { children, ...props } = this.props;
    let newProps = { ...props };
    let { helmetData } = props;

    if (children) {
      newProps = this.mapChildrenToProps(children, newProps);
    }

    if (helmetData && !(helmetData instanceof HelmetData)) {
      const data = helmetData as HelmetDataType;
      helmetData = new HelmetData(data.context, true);
      delete newProps.helmetData;
    }

    return helmetData ? (
      <Dispatcher {...newProps} context={helmetData.value} />
    ) : (
      <Context.Consumer>
        {context => <Dispatcher {...newProps} context={context as DispatcherContextProp} />}
      </Context.Consumer>
    );
  }
}
