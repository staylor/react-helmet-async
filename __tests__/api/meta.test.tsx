import React from 'react';
import type { MockedFunction } from 'vitest';

import { Helmet } from '../../src';
import { HELMET_ATTRIBUTE } from '../../src/constants';
import { render } from '../utils';

Helmet.defaultProps.defer = false;

describe('meta tags', () => {
  describe('API', () => {
    it('updates meta tags', () => {
      render(
        <Helmet
          meta={[
            { charset: 'utf-8' },
            {
              name: 'description',
              content: 'Test description',
            },
            {
              'http-equiv': 'content-type',
              content: 'text/html',
            },
            { property: 'og:type', content: 'article' },
            {
              itemprop: 'name',
              content: 'Test name itemprop',
            },
          ]}
        />
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];

      expect(existingTags).toBeDefined();

      const filteredTags = existingTags.filter(
        tag =>
          tag.getAttribute('charset') === 'utf-8' ||
          (tag.getAttribute('name') === 'description' &&
            tag.getAttribute('content') === 'Test description') ||
          (tag.getAttribute('http-equiv') === 'content-type' &&
            tag.getAttribute('content') === 'text/html') ||
          (tag.getAttribute('itemprop') === 'name' &&
            tag.getAttribute('content') === 'Test name itemprop')
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(4);
    });

    it('clears all meta tags if none are specified', () => {
      render(<Helmet meta={[{ name: 'description', content: 'Test description' }]} />);

      render(<Helmet />);

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags without 'name', 'http-equiv', 'property', 'charset', or 'itemprop' are not accepted", () => {
      render(
        <Helmet
          meta={[
            {
              // @ts-ignore
              href: "won't work",
            },
          ]}
        />
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it('sets meta tags based on deepest nested component', () => {
      render(
        <div>
          <Helmet
            meta={[
              { charset: 'utf-8' },
              {
                name: 'description',
                content: 'Test description',
              },
            ]}
          />
          <Helmet
            meta={[
              {
                name: 'description',
                content: 'Inner description',
              },
              { name: 'keywords', content: 'test,meta,tags' },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag, thirdTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(3);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('charset', 'utf-8');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner description');
      expect(secondTag.outerHTML).toMatchSnapshot();

      expect(thirdTag).toBeInstanceOf(Element);
      expect(thirdTag.getAttribute).toBeDefined();
      expect(thirdTag).toHaveAttribute('name', 'keywords');
      expect(thirdTag).toHaveAttribute('content', 'test,meta,tags');
      expect(thirdTag.outerHTML).toMatchSnapshot();
    });

    it('allows duplicate meta tags if specified in the same component', () => {
      render(
        <Helmet
          meta={[
            {
              name: 'description',
              content: 'Test description',
            },
            {
              name: 'description',
              content: 'Duplicate description',
            },
          ]}
        />
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Test description');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Duplicate description');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('overrides duplicate meta tags with single meta tag in a nested component', () => {
      render(
        <div>
          <Helmet
            meta={[
              {
                name: 'description',
                content: 'Test description',
              },
              {
                name: 'description',
                content: 'Duplicate description',
              },
            ]}
          />
          <Helmet
            meta={[
              {
                name: 'description',
                content: 'Inner description',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it('overrides single meta tag with duplicate meta tags in a nested component', () => {
      render(
        <div>
          <Helmet
            meta={[
              {
                name: 'description',
                content: 'Test description',
              },
            ]}
          />
          <Helmet
            meta={[
              {
                name: 'description',
                content: 'Inner description',
              },
              {
                name: 'description',
                content: 'Inner duplicate description',
              },
            ]}
          />
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner duplicate description');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('does not render tag when primary attribute is null', () => {
      render(
        <Helmet
          meta={[
            {
              name: undefined,
              content: 'Inner duplicate description',
            },
          ]}
        />
      );

      const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toHaveLength(0);
    });

    it('fails gracefully when meta is wrong shape', () => {
      const originalConsole = global.console;
      global.console.warn = vi.fn();

      render(
        <Helmet
          meta={
            // @ts-ignore
            { name: 'title', content: 'some title' }
          }
        />
      );

      const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toHaveLength(0);

      expect(console.warn).toHaveBeenCalled();

      expect((console.warn as MockedFunction<any>).mock.calls[0][0]).toMatchSnapshot();

      global.console = originalConsole;
    });
  });

  describe('Declarative API', () => {
    it('updates meta tags', () => {
      render(
        <Helmet>
          <meta charSet="utf-8" />
          <meta name="description" content="Test description" />
          <meta httpEquiv="content-type" content="text/html" />
          <meta property="og:type" content="article" />
          <meta itemProp="name" content="Test name itemprop" />
        </Helmet>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];

      expect(existingTags).toBeDefined();

      const filteredTags = existingTags.filter(
        tag =>
          tag.getAttribute('charset') === 'utf-8' ||
          (tag.getAttribute('name') === 'description' &&
            tag.getAttribute('content') === 'Test description') ||
          (tag.getAttribute('http-equiv') === 'content-type' &&
            tag.getAttribute('content') === 'text/html') ||
          (tag.getAttribute('itemprop') === 'name' &&
            tag.getAttribute('content') === 'Test name itemprop')
      );

      expect(filteredTags.length).toBeGreaterThanOrEqual(4);
    });

    it('clears all meta tags if none are specified', () => {
      render(
        <Helmet>
          <meta name="description" content="Test description" />
        </Helmet>
      );

      render(<Helmet />);

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it("tags without 'name', 'http-equiv', 'property', 'charset', or 'itemprop' are not accepted", () => {
      render(
        <Helmet>
          <meta
            // @ts-ignore
            href="won't work"
          />
        </Helmet>
      );

      const existingTags = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(0);
    });

    it('sets meta tags based on deepest nested component', () => {
      render(
        <div>
          <Helmet>
            <meta charSet="utf-8" />
            <meta name="description" content="Test description" />
          </Helmet>
          <Helmet>
            <meta name="description" content="Inner description" />
            <meta name="keywords" content="test,meta,tags" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag, thirdTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(3);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('charset', 'utf-8');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner description');
      expect(secondTag.outerHTML).toMatchSnapshot();

      expect(thirdTag).toBeInstanceOf(Element);
      expect(thirdTag.getAttribute).toBeDefined();
      expect(thirdTag).toHaveAttribute('name', 'keywords');
      expect(thirdTag).toHaveAttribute('content', 'test,meta,tags');
      expect(thirdTag.outerHTML).toMatchSnapshot();
    });

    it('allows duplicate meta tags if specified in the same component', () => {
      render(
        <Helmet>
          <meta name="description" content="Test description" />
          <meta name="description" content="Duplicate description" />
        </Helmet>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Test description');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Duplicate description');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('overrides duplicate meta tags with single meta tag in a nested component', () => {
      render(
        <div>
          <Helmet>
            <meta name="description" content="Test description" />
            <meta name="description" content="Duplicate description" />
          </Helmet>
          <Helmet>
            <meta name="description" content="Inner description" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(1);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag.outerHTML).toMatchSnapshot();
    });

    it('overrides single meta tag with duplicate meta tags in a nested component', () => {
      render(
        <div>
          <Helmet>
            <meta name="description" content="Test description" />
          </Helmet>
          <Helmet>
            <meta name="description" content="Inner description" />
            <meta name="description" content="Inner duplicate description" />
          </Helmet>
        </div>
      );

      const existingTags = [...document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`)];
      const [firstTag, secondTag] = existingTags;

      expect(existingTags).toBeDefined();
      expect(existingTags).toHaveLength(2);

      expect(firstTag).toBeInstanceOf(Element);
      expect(firstTag.getAttribute).toBeDefined();
      expect(firstTag).toHaveAttribute('name', 'description');
      expect(firstTag).toHaveAttribute('content', 'Inner description');
      expect(firstTag.outerHTML).toMatchSnapshot();

      expect(secondTag).toBeInstanceOf(Element);
      expect(secondTag.getAttribute).toBeDefined();
      expect(secondTag).toHaveAttribute('name', 'description');
      expect(secondTag).toHaveAttribute('content', 'Inner duplicate description');
      expect(secondTag.outerHTML).toMatchSnapshot();
    });

    it('does not render tag when primary attribute is null', () => {
      render(
        <Helmet>
          <meta name={undefined} content="Inner duplicate description" />
        </Helmet>
      );

      const tagNodes = document.head.querySelectorAll(`meta[${HELMET_ATTRIBUTE}]`);
      const existingTags = [].slice.call(tagNodes);

      expect(existingTags).toHaveLength(0);
    });
  });
});
