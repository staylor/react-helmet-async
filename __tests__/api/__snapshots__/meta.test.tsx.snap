// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`meta tags > API > allows duplicate meta tags if specified in the same component 1`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > allows duplicate meta tags if specified in the same component 2`] = `"<meta name=\\"description\\" content=\\"Duplicate description\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > fails gracefully when meta is wrong shape 1`] = `"Helmet: meta should be of type \\"Array\\". Instead found type \\"object\\""`;

exports[`meta tags > API > overrides duplicate meta tags with single meta tag in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > overrides single meta tag with duplicate meta tags in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > overrides single meta tag with duplicate meta tags in a nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner duplicate description\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > sets meta tags based on deepest nested component 1`] = `"<meta charset=\\"utf-8\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > sets meta tags based on deepest nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`meta tags > API > sets meta tags based on deepest nested component 3`] = `"<meta name=\\"keywords\\" content=\\"test,meta,tags\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > allows duplicate meta tags if specified in the same component 1`] = `"<meta name=\\"description\\" content=\\"Test description\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > allows duplicate meta tags if specified in the same component 2`] = `"<meta name=\\"description\\" content=\\"Duplicate description\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > overrides duplicate meta tags with single meta tag in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > overrides single meta tag with duplicate meta tags in a nested component 1`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > overrides single meta tag with duplicate meta tags in a nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner duplicate description\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > sets meta tags based on deepest nested component 1`] = `"<meta charset=\\"utf-8\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > sets meta tags based on deepest nested component 2`] = `"<meta name=\\"description\\" content=\\"Inner description\\" data-rh=\\"true\\">"`;

exports[`meta tags > Declarative API > sets meta tags based on deepest nested component 3`] = `"<meta name=\\"keywords\\" content=\\"test,meta,tags\\" data-rh=\\"true\\">"`;
